README.md
# ruby-mp3info

* http://github.com/moumar/ruby-mp3info

[![Build Status](https://travis-ci.org/moumar/ruby-mp3info.png?branch=master)](https://travis-ci.org/moumar/ruby-mp3info)

## Description

ruby-mp3info read low-level informations and manipulate tags on mp3 files.

## Features/Problems

* Written in pure ruby
* Read low-level informations like bitrate, length, samplerate, etc...
* Read, write, remove id3v1 and id3v2 tags
* Correctly read VBR files (with or without Xing header)
* Only 2.3 version is supported for writings id3v2 tags
* id3v2 tags are always written in UTF-16 encoding

## Synopsis

```ruby
require "mp3info"

# Read and display infos & tags

Mp3Info.open("myfile.mp3") do |mp3info|
  puts mp3info
end

# read/write tag1 and tag2 with Mp3Info#tag attribute
# when reading tag2 have priority over tag1
# when writing, each tag is written.

Mp3Info.open("myfile.mp3") do |mp3|
  puts mp3.tag.title
  puts mp3.tag.artist
  puts mp3.tag.album
  puts mp3.tag.tracknum
  mp3.tag.title = "track title"
  mp3.tag.artist = "artist name"
end

# tags are written when calling Mp3Info#close or at the end of the #open block

### access id3v2 tags

Mp3Info.open("myfile.mp3") do |mp3|
  # you can access four letter v2 tags like this
  puts mp3.tag2.TIT2
  mp3.tag2.TIT2 = "new TIT2"
  # or like that
  mp3.tag2["TIT2"]
  # at this time, only COMM tag is processed after reading and before writing
  # according to ID3v2#options hash
  mp3.tag2.options[:lang] = "FRE"
  mp3.tag2.COMM = "my comment in french, correctly handled when reading and writing"
end

### image manipulation

file = File.new('input_img','rb')
Mp3Info.open '1.mp3' do |m|
   pictures = m.tag2.pictures # array of images :
   pictures.each do |description, data|
      # description ends with (.jpg / .png) for easy writing to file
      File.binwrite(description, data)
  end
  m.tag2.remove_pictures # remove all images

  m.tag2.add_picture(file.read) # add a single image to APIC tag
  # **optionally**, you may include options for the image tag to add_picture():
  # options = { mime: 'jpeg|gif|png', description: "description",
  #             pic_type: 0 }
  # There is no need to specify mime or options because the mime is guessed based on the input image
  # Only one image is permitted in the mp3 for compatibility. add_picture() overwrites all previous images.
  # Please note:
  # If supplying :mime option just supply the mime 'jpeg', 'gif' or 'png', the code adds the "image/.." for you!
  # e.g. m.tag2.add_picture(file.read, mime: 'jpeg', pic_type: 3) gets a mime "image/jpeg"
end
```

## Requirements

* iconv when using ruby 1.8

## Install

    gem install ruby-mp3info

## Developers

After checking out the source, run:

    $ rake newb

This task will install any missing dependencies, run the tests/specs, and generate the RDoc.

## License

GPL-3.0

## TODO:

* encoder detection
* support for more tags in id3v2
* generalize id3v2 with other audio formats (APE, MPC, OGG, etc...)


History.txt
=== 0.8.10 / 2016-10-31

* bugfix on /^(T|COM|USLT)/ tags parsing

=== 0.8.9 / 2016-02-03

* bugfix on JPEG picture parsing + testsuite

=== 0.8.8 / 2016-01-26

* fix on picture parsing when mime type is wrong

=== 0.8.7 / 2015-07-13

*  Cope with malformed UTF16 string tags without BOM  (thanks to mikecrowe)
*  Fixes for when the desc + data are not unpacked correctly for an id3v2 jpeg (thanks to ekaminsky)

=== 0.8.6 / 2015-03-28

* fixed frame size computation (thanks to ekaminski)
* now correctly close the @io (thanks to cyberroot)
* correctly compute length of VBR files (fixes #42)

=== 0.8.5 / 2014-09-17

* lyrics support (USLT tag) (thanks to ipmsteven)
* fixed travis CI tests (thanks to ipmsteven)

=== 0.8.4 / 2014-04-26

* more robust frame scanning when unsynced (thanks to emonsqueeze for the bug report)

=== 0.8.3 / 2014-01-18

* fix for ruby 2.1.0 (thanks to smashwilson)

=== 0.8.2 / 2013-11-20

* better parsing of corrupted mp3 (more robust frame scanning)
* upgraded license to GPL 3 (previous license (ruby) is compatible)

=== 0.8.1 / 2013-10-28

* test fixes
* added dot to temp filename

=== 0.8 / 2013-01-28

* allow higher level of reading and writing to APIC tag:
* tag2.pictures, tag2.remove_pictures, tag2.add_picture
* added "<<...snip...>>>" to shorten APIC tag when Hash.inspect called (not a shallow copy hash, deep copied hash so does not effect data) tag2.inspect

=== 0.7.2 / 2012-12-24

* do not try to commit changes to IO or StringIO

=== 0.7.1 / 2012-04-01

* convert to utf-16 little endian (instead of big) for more compatibility

=== 0.7 / 2012-02-29

* removed iconv for ruby >= 1.9
* default encoding for decoded and encoded tags is UTF-8 now (removed :encoding params for ID3v2)
* come back to a non senseless version numbering

=== 0.6.16 / 2011-11-10

* fixed type error when inspecting mp3info (thanks to Jacob Lichner)

=== 0.6.15 / 2011-07-18

* support for StringIO as input (thanks to Edd Parris)

=== 0.6.14 / 2011-06-17

* Added a check for nil that was seen causing problems when processing files. (thanks to Carl Hall)
* Fixed reading on win32, requires binary flag. (thanks to Jonas Tingeborn)
* Fixed white spaces. Replaced tabs with spaces to make the source readable on for users other than the original author. (thanks to Jonas Tingeborn)
* Add :parse_mp3 flag to new/open. (thanks to Dave Lee)
* Add benchmark for parsing performance. (thanks to Dave Lee)
* fixed ID3v2#io_position computing, so Mp3Info#audio_content() is correct now

=== 0.6.13 / 2009-05-26

* fixed bad mapping of artist inside id3 2.2
* adding fusil fuzzer tests
* Improved support for id3v2.2 to id3v2.3 field mapping
* each_frame() iterator
* removed @bitrate & @length computation based on @tag2[TLEN]

=== 0.6.12 / 2009-02-23

* fixed bug when @tag2["TLEN"] == 0

=== 0.6.11 / 2009-01-27

* the library doesn't raise an ID3v2Error anymore when a id3v2 tag size is incorrect, but just output
  a warning. (Fixes bug #23619)

=== 0.6.10 / 2008-11-27

* processing of tags (read and write) can be disabled with :parse_tags => false

=== 0.6.9 / 2008-09-16

* now correctly remove trailing whitespaces form tag1 values
* bugfix #21687 included: 'tweaked the MP3 frame synch code to parse certain mp3 files'

=== 0.6.8 / 2008-08-20

* support for MPEG 2.5 (thanks to Oleguer Huguet Ibars)
* support for vbr files without Xing header

=== 0.6.7 / 2008-06-26

* Mp3Info#header hash now gives access to additional mpeg attributes (thanks to Andrew Kuklewicz)

=== 0.6.6 / 2008-05-27

* avoid reading tag that are too big (> 50Mb)
* ruby 1.9 support (thanks to Dave Thomas)
* FIXED: bug #20311 'Multiple APIC frames may be stored incorrectly'
* FIXED: bug #20312 'doesn't use v2.2 frames for extracting meta data'

=== 0.6.5 / 2008-04-19

* added Mp3Info#audio_content method, to return "audio-only" boundaries from mp3, i.e. data without tags
  (closes feature request #17230)
* bugfix on reading id3 2.4 (size not syncsafed)
* more robust tag decoding with bad tags

=== 0.6.4 / 2008-04-16

*  added @tag2["disc_number"] and @tag2["disc_total"] mirroring TPOS attribute (thanks to Harry Ohlsen)

=== 0.6.3 / 2008-03-28

* some internals modifications for the compatibility with ruby-audioinfo

=== 0.6.2 / 2008-03-02

* better handling of frames:
  decode and encode as raw string by default,
  or handle charset decoding/encoding for /^T/ and
  COMM frames

=== 0.6.1 / 2008-02-28

* FIXED: fails to read id3v2 tags when iconv fails

=== 0.6 / 2008-02-24

* FIXED: correct handling of encoding in id3v2 tags

=== 0.5.1 / 2007-09-10

* ADDED: Mp3Info#reload method to reload the file from the disk
* FIXED: bug [#2604] Not able to delete tag1
* FIXED: bug #3401 'id3v2.rb dies when trying to read a certain mp3'
* FIXED: bug #2957 'Error message "Can't define singleton"'
* FIXED: bug #3068 'require_gem ("ruby-mp3info") doesn't works'
* FIXED: bug #11967 "Leading 'h' from 'http://' gets chopped on URL fields"
* PATCHED: with patch #3157 'Fix for 64 bit Ruby'

=== 0.5 / 2005-12-06

* id3v2 writing and removing support added. tag2 attribute is r/w now
* max guess size to find a valid frame set to 2Mb
* implemented a new class ID3v2, ID2TAGS moved into it
* Mp3Info.tag is r/w now and has priority over @tag1 and @tag2 when writing
* added Mp3Info#rename() method to change the filename written at close
* clean up: all overloaded standards classes replaced by including modules
* FIXED bug in reading id3v2 tags tagged with olds versions of "mp3ext" ( http://www.mutschler.de/mp3ext/ )
* FIXED bug on calculating id3v2 frame size
* FIXED bug when multiple TLEN tags

lib/mp3info/id3v2.rb
# encoding: utf-8
# License:: Ruby
# Author:: Guillaume Pierronnet (mailto:guillaume.pierronnet@gmail.com)

require "delegate"

require "mp3info/extension_modules"

class ID3v2Error < StandardError ; end

# This class can be used to decode id3v2 tags from files, like .mp3 or .ape for example.
# It works like a hash, where key represents the tag name as 3 or 4 upper case letters
# (respectively related to 2.2 and 2.3+ tag) and value represented as array or raw value.
# Written version is always 2.3.
class ID3v2 < DelegateClass(Hash)

  TAGS = {
    "AENC" => "Audio encryption",
    "APIC" => "Attached picture",
    "COMM" => "Comments",
    "COMR" => "Commercial frame",
    "ENCR" => "Encryption method registration",
    "EQUA" => "Equalization",
    "ETCO" => "Event timing codes",
    "GEOB" => "General encapsulated object",
    "GRID" => "Group identification registration",
    "IPLS" => "Involved people list",
    "LINK" => "Linked information",
    "MCDI" => "Music CD identifier",
    "MLLT" => "MPEG location lookup table",
    "OWNE" => "Ownership frame",
    "PRIV" => "Private frame",
    "PCNT" => "Play counter",
    "POPM" => "Popularimeter",
    "POSS" => "Position synchronisation frame",
    "RBUF" => "Recommended buffer size",
    "RVAD" => "Relative volume adjustment",
    "RVRB" => "Reverb",
    "SYLT" => "Synchronized lyric/text",
    "SYTC" => "Synchronized tempo codes",
    "TALB" => "Album/Movie/Show title",
    "TBPM" => "BPM (beats per minute)",
    "TCOM" => "Composer",
    "TCON" => "Content type",
    "TCOP" => "Copyright message",
    "TDAT" => "Date",
    "TDLY" => "Playlist delay",
    "TENC" => "Encoded by",
    "TEXT" => "Lyricist/Text writer",
    "TFLT" => "File type",
    "TIME" => "Time",
    "TIT1" => "Content group description",
    "TIT2" => "Title/songname/content description",
    "TIT3" => "Subtitle/Description refinement",
    "TKEY" => "Initial key",
    "TLAN" => "Language(s)",
    "TLEN" => "Length",
    "TMED" => "Media type",
    "TOAL" => "Original album/movie/show title",
    "TOFN" => "Original filename",
    "TOLY" => "Original lyricist(s)/text writer(s)",
    "TOPE" => "Original artist(s)/performer(s)",
    "TORY" => "Original release year",
    "TOWN" => "File owner/licensee",
    "TPE1" => "Lead performer(s)/Soloist(s)",
    "TPE2" => "Band/orchestra/accompaniment",
    "TPE3" => "Conductor/performer refinement",
    "TPE4" => "Interpreted, remixed, or otherwise modified by",
    "TPOS" => "Part of a set",
    "TPUB" => "Publisher",
    "TRCK" => "Track number/Position in set",
    "TRDA" => "Recording dates",
    "TRSN" => "Internet radio station name",
    "TRSO" => "Internet radio station owner",
    "TSIZ" => "Size",
    "TSRC" => "ISRC (international standard recording code)",
    "TSSE" => "Software/Hardware and settings used for encoding",
    "TYER" => "Year",
    "TXXX" => "User defined text information frame",
    "UFID" => "Unique file identifier",
    "USER" => "Terms of use",
    "USLT" => "Unsychronized lyric/text transcription",
    "WCOM" => "Commercial information",
    "WCOP" => "Copyright/Legal information",
    "WOAF" => "Official audio file webpage",
    "WOAR" => "Official artist/performer webpage",
    "WOAS" => "Official audio source webpage",
    "WORS" => "Official internet radio station homepage",
    "WPAY" => "Payment",
    "WPUB" => "Publishers official webpage",
    "WXXX" => "User defined URL link frame"
  }

  # Translate V2 to V3 tags
  TAG_MAPPING_2_2_to_2_3 = {
    "BUF"   => "RBUF",
    "COM"   => "COMM",
    "CRA"   => "AENC",
    "EQU"   => "EQUA",
    "ETC"   => "ETCO",
    "GEO"   => "GEOB",
    "MCI"   => "MCDI",
    "MLL"   => "MLLT",
    "PIC"   => "APIC",
    "POP"   => "POPM",
    "REV"   => "RVRB",
    "RVA"   => "RVAD",
    "SLT"   => "SYLT",
    "STC"   => "SYTC",
    "TAL"   => "TALB",
    "TBP"   => "TBPM",
    "TCM"   => "TCOM",
    "TCO"   => "TCON",
    "TCR"   => "TCOP",
    "TDA"   => "TDAT",
    "TDY"   => "TDLY",
    "TEN"   => "TENC",
    "TFT"   => "TFLT",
    "TIM"   => "TIME",
    "TKE"   => "TKEY",
    "TLA"   => "TLAN",
    "TLE"   => "TLEN",
    "TMT"   => "TMED",
    "TOA"   => "TOPE",
    "TOF"   => "TOFN",
    "TOL"   => "TOLY",
    "TOR"   => "TORY",
    "TOT"   => "TOAL",
    "TP1"   => "TPE1",
    "TP2"   => "TPE2",
    "TP3"   => "TPE3",
    "TP4"   => "TPE4",
    "TPA"   => "TPOS",
    "TPB"   => "TPUB",
    "TRC"   => "TSRC",
    "TRD"   => "TRDA",
    "TRK"   => "TRCK",
    "TSI"   => "TSIZ",
    "TSS"   => "TSSE",
    "TT1"   => "TIT1",
    "TT2"   => "TIT2",
    "TT3"   => "TIT3",
    "TXT"   => "TEXT",
    "TXX"   => "TXXX",
    "TYE"   => "TYER",
    "UFI"   => "UFID",
    "ULT"   => "USLT",
    "WAF"   => "WOAF",
    "WAR"   => "WOAR",
    "WAS"   => "WOAS",
    "WCM"   => "WCOM",
    "WCP"   => "WCOP",
    "WPB"   => "WPB",
    "WXX"   => "WXXX",
  }

  # See id3v2.4.0-structure document, at section 4.
  TEXT_ENCODINGS = ["iso-8859-1", "utf-16", "utf-16be", "utf-8"]

  include Mp3Info::HashKeys

  # In Ruby 2.1.0 (and possibly others), DelegateClass breaks Kernel methods.
  include Kernel

  # this is the position in the file where the tag really ends
  attr_reader :io_position

  # :+lang+: for writing comments
  #
  # [DEPRECATION] :+encoding+: one of the string of +TEXT_ENCODINGS+,
  # use of :encoding parameter is DEPRECATED. In ruby 1.8, use utf-8 encoded strings for tags.
  # In ruby >= 1.9, strings are automatically transcoded from their originaloriginal  encoding.
  attr_reader :options

  # possible options are described above ('options' attribute)
  # you can access this object like an hash, with [] and []= methods
  # special cases are ["disc_number"] and ["disc_total"] mirroring TPOS attribute
  def initialize(options = {})
    @options = { :lang => "ENG" }
    if @options[:encoding]
      warn("use of :encoding parameter is DEPRECATED. In ruby 1.8, use utf-8 encoded strings for tags.\n" +
           "In ruby >= 1.9, strings are automatically transcoded from their original encoding.")
    end

    @options.update(options)

    @hash = {}
    #TAGS.keys.each { |k| @hash[k] = nil }
    @hash_orig = {}
    super(@hash)
    @parsed = false
    @version_maj = @version_min = nil
  end

  # does this tag has been correctly read ?
  def parsed?
    @parsed
  end

  # does this tag has been changed ?
  def changed?
    @hash_orig != @hash
  end

  # full version of this tag (like "2.3.0") or nil
  # if tag was not correctly read
  def version
    if @version_maj && @version_min
      "2.#{@version_maj}.#{@version_min}"
    else
      nil
    end
  end

  ### cuts out long tag values from hash for display on screen
  def to_inspect_hash
    result = Marshal.load(Marshal.dump(self.to_hash))
    result.each do |k,v|
      if v.is_a? Array
        v.each_index do |i, item|
          if (v[i].is_a? String and v[i].length > 128)
            result[k][i] = pretty_header(v[i])
          end
        end
      elsif v.is_a? String and v.length > 128
        result[k] = pretty_header(v) # this method 'snips' long data
      end
    end
    result
  end

  ### ID3V2::add_picture
  ### Takes an image string as input and writes it with header.
  ### Mime type is automatically guessed by default.
  ### It is possible but not necessary to include:
  ###  :pic_type => 0 - 14 (see http://id3.org/id3v2.3.0#Attached_picture)
  ###  :mime => 'gif'
  ###  :description => "Image description"
  def add_picture(data, opts = {})
    options = {
                :pic_type => 0,
                :mime => nil,
                :description => "image"
              }
    options.update(opts)
    jpg = Regexp.new( "^\xFF".force_encoding("BINARY"),
                 Regexp::FIXEDENCODING )
    png = Regexp.new( "^\x89PNG".force_encoding("BINARY"),
                 Regexp::FIXEDENCODING )
    gif = Regexp.new( "^\x89GIF".force_encoding("BINARY"),
                 Regexp::FIXEDENCODING )

    mime = options[:mime]
    mime ||= "jpg" if data.match jpg
    mime ||= "png" if data.match png
    mime ||= "gif" if data.match gif
    pic_type = options[:pic_type]
    pic_type = ["%02i" % pic_type].pack('H*')
    desc = "#{options[:description]}"
    header = "\x00image/#{mime}\x00#{pic_type}#{desc}\x00"
    self["APIC"] = header + data.force_encoding('BINARY')
  end

  ### Returns an array of images:
  ###   [  ["01_.jpg", "Image Data in Binary String"],
  ###      ["02_.png", "Another Image in a String"]    ]
  ###
  ### e.g. to write all images:
  ### mp3.tag2.pictures.each do |image|
  ###   File.open(img[0], 'wb'){|f| f.write img[1])
  ### end
  def pictures
    apic_images = [self["APIC"]].flatten.dup
    result = []
    apic_images.each_index do |index|
      pic = apic_images[index]
      next if !pic.is_a?(String) or pic == ""
      pic.force_encoding 'BINARY'
      picture = []
      jpg_regexp = Regexp.new("jpg|JPG|jpeg|JPEG".force_encoding("BINARY"),
                   Regexp::FIXEDENCODING )
      png_regexp = Regexp.new("png|PNG".force_encoding("BINARY"),
                   Regexp::FIXEDENCODING )
      header = pic.unpack('a120').first.force_encoding "BINARY"
      mime_pos = 0

      # safest way to correctly extract jpg and png is finding mime
      if header.match jpg_regexp and not header.match png_regexp
        mime = "jpg"
        mime_pos = header =~ jpg_regexp
        start = Regexp.new("\xFF\xD8".force_encoding("BINARY"),
                Regexp::FIXEDENCODING )
        start_with_anchor = Regexp.new("^\xFF\xD8".force_encoding("BINARY"),
                                       Regexp::FIXEDENCODING )
      end

      if header.match png_regexp and not header.match jpg_regexp
        mime = "png"
        mime_pos = header =~ png_regexp
        start = Regexp.new("\x89PNG".force_encoding("BINARY"),
                Regexp::FIXEDENCODING )
        start_with_anchor = Regexp.new("^\x89PNG".force_encoding("BINARY"),
                            Regexp::FIXEDENCODING )
      end

      puts "analysing image: #{header.inspect}..." if $DEBUG
      _, _, desc, data = pic[mime_pos, pic.length].unpack('Z*hZ*a*')

      if mime
        if !data.match(start_with_anchor) or data.nil?
          real_start = pic =~ start
          data = pic[real_start, pic.length]
        end

        if mime == "jpg"
          # inspect jpg image header (first 10 chars) for "\xFF\x00" (expect "\xFF")
          trailing_null_byte = Regexp.new("(\377)(\000)".force_encoding('BINARY'),
                                          Regexp::FIXEDENCODING)
          md = data =~ trailing_null_byte
          if !md.nil? and md < 10
            data.gsub!(trailing_null_byte, "\xff".force_encoding('BINARY'))
          end
        end
      else
        mime = "dat"
        # if no jpg or png, extract data anyway e.g. gif
        mime, desc, data = pic.unpack('h Z* h Z* a*').values_at(1,3,4)
      end

      filename = ("%02i_#{desc[0,25]}" % (index + 1)).gsub('/','')

      picture[0] = filename
      picture[1] = data
      result << picture
    end
    result
  end

  def inspect
    self.to_inspect_hash
  end

  def remove_pictures
    self["APIC"] = ""
    self["PIC"]  = ""
  end

  ### gets id3v2 tag information from io object (must support #seek() method)
  def from_io(io)
    @io = io
    original_pos = @io.pos
    @io.extend(Mp3Info::Mp3FileMethods)
    version_maj, version_min, flags = @io.read(3).unpack("CCB4")
    @unsync, ext_header, _, _ = (0..3).collect { |i| flags[i].chr == '1' }  # _, _ = experimental, footer
    raise(ID3v2Error, "can't find version_maj ('#{version_maj}')") unless [2, 3, 4].include?(version_maj)
    @version_maj, @version_min = version_maj, version_min
    @tag_length = @io.get_syncsafe

    @parsed = true
    begin
      case @version_maj
        when 2
          read_id3v2_2_frames
        when 3, 4
          # seek past extended header if present
          @io.seek(@io.get_syncsafe - 4, IO::SEEK_CUR) if ext_header
          read_id3v2_3_frames
      end
    rescue ID3v2Error => e
      warn("warning: id3v2 tag not fully parsed: #{e.message}")
    end
    @io_position = @io.pos
    @tag_length = @io_position - original_pos

    @hash_orig = @hash.dup
    #no more reading
    @io = nil
  end

  # dump tag for writing. Version is always 2.3.0
  def to_bin
    #TODO handle of @tag2[TLEN"]
    #TODO add of crc
    #TODO add restrictions tag

    tag = ""
    @hash.each do |k, v|
      next unless v
      next if v.respond_to?("empty?") and v.empty?

      # Automagically translate V2 to V3 tags
      k = TAG_MAPPING_2_2_to_2_3[k] if TAG_MAPPING_2_2_to_2_3.has_key?(k)

      # doesn't encode id3v2.2 tags, which have 3 characters
      next if k.size != 4

      # Output one flag for each array element, or one only if it's not an array
      [v].flatten.each do |value|
        data = encode_tag(k, value.to_s)
        #data << "\x00"*2 #End of tag

        tag << k[0,4]   #4 characte max for a tag's key
        #tag << to_syncsafe(data.size) #+1 because of the language encoding byte
        size = data.size
        size = data.dup.force_encoding("binary").size
        tag << [size].pack("N") #+1 because of the language encoding byte
        tag << "\x00"*2 #flags
        tag << data
      end
    end

    tag_str = "ID3"
    #version_maj, version_min, unsync, ext_header, experimental, footer
    tag_str << [ 3, 0, "0000" ].pack("CCB4")
    tag_str << [to_syncsafe(tag.size)].pack("N")
    tag_str << tag
    puts "tag in binary format: #{tag_str.inspect}" if $DEBUG
    tag_str
  end

  private

  def encode_tag(name, value)
    puts "encode_tag(#{name.inspect}, #{value.inspect})" if $DEBUG
    name = name.to_s

    if name =~ /^(COM|T|USLT)/
      transcoded_value = Mp3Info::EncodingHelper.convert_to(value, "utf-8", "utf-16")
    end
    case name
    when "COMM", "USLT"
      puts "encode #{name} lang: #{@options[:lang]}, value #{transcoded_value.inspect}" if $DEBUG
      s = [ 1, @options[:lang], "\xFE\xFF\x00\x00", transcoded_value].pack("ca3a*a*")
      return s
    when /^T/
      transcoded_value.force_encoding("BINARY")

	return "\x01" + transcoded_value
      else
        return value
    end
  end

  ### Read a tag from file and perform UNICODE translation if needed
  def decode_tag(name, raw_value)
    puts("decode_tag(#{name.inspect}, #{raw_value.inspect})") if $DEBUG
    if name =~ /^(T|COM|USLT)/
      begin
        if name =~ /^(COM|USLT)/
          encoding_index, lang, raw_tag = raw_value.unpack("ca3a*")
          if raw_tag == "\x00"
            return nil
          else
            if encoding_index == 1
              split_str = /\x00\x00\x00?/
            else
              split_str = "\x00"
            end
            head, tail = raw_tag.split(split_str)
            if head == "\xFF\xFE".force_encoding('binary')
              rgx = Regexp.new("^\xFF\xFE\x00\x00".force_encoding("BINARY"))
              out = raw_tag.sub(rgx, '')
            elsif tail
              out = tail
            elsif head
              out = head
            else
              warn("warning: cannot decode tag #{name} with raw value #{raw_value.inspect}")
              return nil
            end
          end
          puts "COM tag found. encoding: #{encoding_index} lang: #{lang} str: #{out.inspect}" if $DEBUG
        else
          encoding_index = raw_value.getbyte(0) # language encoding (see TEXT_ENCODINGS constant)
          out = raw_value[1..-1]
        end
        # we need to convert the string in order to match
        # the requested encoding
        if TEXT_ENCODINGS[encoding_index]
          if encoding_index == 1
            out = Mp3Info::EncodingHelper.decode_utf16(out)
          else
            out.force_encoding(TEXT_ENCODINGS[encoding_index])
          end
          out.encode!("utf-8")
        end

        # remove padding zeros for textual tags
        r = Regexp.new("\x00*$".encode(out.encoding))
        out.sub!(r, '')

        return out
      rescue => e
        warn "warning: cannot decode tag #{name} with raw value #{raw_value.inspect}: #{e}"
        return nil
      end
    else
      return raw_value
    end
  end

  ### reads id3 ver 2.3.x/2.4.x frames and adds the contents to @tag2 hash
  ### NOTE: the id3v2 header does not take padding zero's into consideration
  def read_id3v2_3_frames
    loop do # there are 2 ways to end the loop
      name = @io.read(4)
      if name.nil? || name.getbyte(0) == 0 || name == "MP3e" #bug caused by old tagging application "mp3ext" ( http://www.mutschler.de/mp3ext/ )
        @io.seek(-4, IO::SEEK_CUR)    # 1. find a padding zero,
	seek_to_v2_end
        break
      else
	if @version_maj == 4
	  size = @io.get_syncsafe
	else
	  size = @io.get32bits
	end
        @io.seek(2, IO::SEEK_CUR)     # skip flags
        puts "name '#{name}' size #{size}" if $DEBUG
        add_value_to_tag2(name, size)
      end
      break if @io.pos >= @tag_length # 2. reach length from header
    end
  end

  ### reads id3 ver 2.2.x frames and adds the contents to @tag2 hash
  ### NOTE: the id3v2 header does not take padding zero's into consideration
  def read_id3v2_2_frames
    loop do
      name = @io.read(3)
      if name.nil? || name.getbyte(0) == 0
        @io.seek(-3, IO::SEEK_CUR)
	seek_to_v2_end
        break
      else
        size = (@io.getbyte << 16) + (@io.getbyte << 8) + @io.getbyte
	add_value_to_tag2(name, size)
        break if @io.pos >= @tag_length
      end
    end
  end

  ### Add data to tag2["name"]
  ### read lang_encoding, decode data if unicode and
  ### create an array if the key already exists in the tag
  def add_value_to_tag2(name, size)
    puts "add_value_to_tag2" if $DEBUG

    if size > 50_000_000
      raise ID3v2Error, "tag size is > 50_000_000"
    end

    data_io = @io.read(size)
    data = decode_tag(name, data_io)
    if data && !data.empty?
      if self.keys.include?(name)
        if self[name].is_a?(Array)
          unless self[name].include?(data)
            self[name] << data
          end
        else
          self[name] = [ self[name], data ]
        end
      else
        self[name] = data
      end

      if name == "TPOS" && data =~ /(\d+)\s*\/\s*(\d+)/
        self["disc_number"] = $1.to_i
        self["disc_total"] = $2.to_i
      end
    end

    puts "self[#{name.inspect}] = #{self[name].inspect}" if $DEBUG
  end

  ### runs thru @file one char at a time looking for best guess of first MPEG
  ###  frame, which should be first 0xff byte after id3v2 padding zero's
  def seek_to_v2_end
    until @io.getbyte == 0xff
      raise ID3v2Error, "got EOF before finding id3v2 end" if @io.eof?
    end
    @io.seek(-1, IO::SEEK_CUR)
  end

  ### convert an 32 integer to a syncsafe string
  def to_syncsafe(num)
    ( (num<<3) & 0x7f000000 )  + ( (num<<2) & 0x7f0000 ) + ( (num<<1) & 0x7f00 ) + ( num & 0x7f )
  end

  ### this is especially useful for printing out APIC data because
  ### only the header of the APIC tag is of interest
  def pretty_header(str, chars=128)
    "#{str.unpack("a#{chars}").first}<<<...snip...>>>".inspect[1..-2]
  end

end
